cql3/restrictions/statement_restrictions.cc

```diff
@@ -598,14 +600,93 @@ void statement_restrictions::process_clustering_columns_restrictions(bool for_vi
+    return dht::partition_range::make_singular(std::move(pos));
+}
+
+void error_if_exceeds(size_t size, size_t limit) {
+    if (size > limit) {
+        throw std::runtime_error(
+                fmt::format("clustering-key cartesian product size {} is greater than maximum {}", size, limit));
+    }
+}
+
+/// Computes partition-key ranges from expressions, which contains EQ/IN for every partition column.
+dht::partition_range_vector partition_ranges_from_singles(
+        const std::vector<expr::expression>& expressions, const query_options& options, const schema& schema) {
+    const size_t size_limit =
+            options.get_cql_config().restrictions.partition_key_restrictions_max_cartesian_product_size;
+    // Each element is a vector of that column's possible values:
+    std::vector<std::vector<bytes>> column_values(schema.partition_key_size());
+    size_t product_size = 1;
+    for (const auto& e : expressions) {
+        if (const auto arbitrary_binop = find_atom(e, [] (const binary_operator&) { return true; })) {
+            if (auto cv = std::get_if<expr::column_value>(&arbitrary_binop->lhs)) {
+                const value_set vals = possible_lhs_values(cv->col, e, options);
+                if (auto lst = std::get_if<value_list>(&vals)) {
+                    if (lst->empty()) {
+                        return {};
+                    }
+                    product_size *= lst->size();
+                    error_if_exceeds(product_size, size_limit);
+                    column_values[schema.position(*cv->col)] = *lst;
+                } else {
+                    throw invalid_request("use token!");
+                }
+            }
+        }
+    }
+    cartesian_product cp(column_values);
+    dht::partition_range_vector ranges(cartesian_product_size(column_values));
+    std::transform(cp.begin(), cp.end(), ranges.begin(), std::bind_front(range_from_bytes, schema));
+    return ranges;
+}
+
+} // anonymous namespace
+
 dht::partition_range_vector statement_restrictions::get_partition_key_ranges(const query_options& options) const {
-    if (_partition_key_restrictions->empty()) {
+    if (_partition_range_restrictions.empty()) {
         return {dht::partition_range::make_open_ended_both_sides()};
     }
-    if (_partition_key_restrictions->needs_filtering(*_schema)) {
-        return {dht::partition_range::make_open_ended_both_sides()};
+    if (has_token(_partition_range_restrictions[0])) {
+        if (_partition_range_restrictions.size() != 1) {
+            on_internal_error(
+                    rlogger,
+                    format("Unexpected size of token restrictions: {}", _partition_range_restrictions.size()));
+        }
+        return partition_ranges_from_token(_partition_range_restrictions[0], options);
     }
-    return _partition_key_restrictions->bounds_ranges(options);
+    return partition_ranges_from_singles(_partition_range_restrictions, options, *_schema);
```

Comment:
This uses a common path for cartesian products and a plain partition key equality. It's elegant in that it avoids special cases, but if we have a performance regression we can consider special-casing the common case of a cartesian product of size 1.

Reply:
I tried special-casing PK equality, but the performance didn't budge.  The cartesian_product wrapper seems very efficient.